Shared/Tests.Shared/AddOrUpdateTests.cs

```diff
@@ -846,6 +846,86 @@ public void AddOrUpdate_WhenObjectHasNonEmptyList_ShouldNotThrow()
             }, Throws.Nothing);
         }
 
+        [Test]
+        public void AddOrUpdate_WhenListIsNull_ShouldNotThrow()
+        {
+            var first = new PrimaryKeyWithPKList
+            {
+                Id = 42
+            };
+
+            Assert.That(() =>
+            {
+                _realm.Write(() => _realm.Add(first, update: true));
+            }, Throws.Nothing);
+        }
+
+        [Test]
+        public void AddOrUpdate_WhenListObjectsHavePK_ShouldOverwriteList()
+        {
+            var first = new PrimaryKeyWithPKList
+            {
+                Id = 1
+            };
+
+            first.ListValue.Add(new PrimaryKeyObject
+            {
+                Id = 1
+            });
+
+            _realm.Write(() => _realm.Add(first));
+
+            var second = new PrimaryKeyWithPKList
+            {
+                Id = 1
+            };
+
+            second.ListValue.Add(new PrimaryKeyObject
+            {
+                Id = 2
+            });
+
+            _realm.Write(() => _realm.Add(second, update: true));
+
+            Assert.That(first.ListValue, Is.EquivalentTo(second.ListValue));
+            Assert.That(first.ListValue.Count, Is.EqualTo(1));
+            Assert.That(first.ListValue[0].Id, Is.EqualTo(2));
+            Assert.That(_realm.All<PrimaryKeyObject>().Count(), Is.EqualTo(2));
+        }
+
+        [Test]
+        public void AddOrUpdate_WhenListObjectsDontHavePK_ShouldOverwriteList()
+        {
+            var first = new PrimaryKeyWithNoPKList
+            {
+                Id = 1
+            };
+
+            first.ListValue.Add(new NonPrimaryKeyObject
+            {
+                StringValue = "1"
+            });
+
+            _realm.Write(() => _realm.Add(first));
+
+            var second = new PrimaryKeyWithNoPKList
+            {
+                Id = 1
+            };
+
+            second.ListValue.Add(new NonPrimaryKeyObject
+            {
+                StringValue = "2"
+            });
+
+            _realm.Write(() => _realm.Add(second, update: true));
+
+            Assert.That(first.ListValue, Is.EquivalentTo(second.ListValue));
+            Assert.That(first.ListValue.Count, Is.EqualTo(1));
+            Assert.That(first.ListValue[0].StringValue, Is.EqualTo("2"));
+            Assert.That(_realm.All<NonPrimaryKeyObject>().Count(), Is.EqualTo(2));
```

Comment:
The original exception was reproduced by a test like https://github.com/realm/realm-dotnet/pull/1040/files/f9ee927531099a1885cb6329c21fdae5e8081457#diff-601e2ac442a9e2a156fbd52803ca8dc1R850, but I could add a clearing test just in case.